aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -240,6 +241,13 @@ private void newStream() throws IOException {
 
   @Override
   public void close() throws IOException {
+
+    // A failed s3 close removes state that is required for a successful close.
+    if (closeFailureException != null) {
+      throw new IOException(
+          "Attempted to close a S3 stream that failed to close earlier", closeFailureException);
```

Comment:
@rdblue We had a discussion about this offline and I think this is probably the only safe thing to do. While double-closing shouldn't throw an exception, if the close failed that means that any future close is also a failure. The S3OutputStream loses it's internal state on close which means that a subsequent close call can never succeed if the IOException is handled.

Because we can't ever retry successfully we decided that the stream should keep throwing exceptions to allow a caller to not falsely believe that a retry close had succeeded when the original close failed.

Reply:
This makes sense to me.

Nit: the error msg doesn't read smoothly. what about `Attempted to close a S3 output stream that has earlier close failed already`.

Reply:
Thanks for the suggestion @stevenzwu, made few changes based on the suggestion.